addons/l10n_es_edi_sii/models/account_edi_format.py

```diff
@@ -0,0 +1,603 @@
+                    + tax_details_info_other_vals['tax_amount_deductible']
+                ), 2)
+
+            info_list.append(info)
+        return info_list
+
+    def _l10n_es_edi_web_service_aeat_vals(self, invoices):
+        if invoices[0].is_sale_document():
+            return {'url': 'https://www2.agenciatributaria.gob.es/static_files/common/internet/dep/aplicaciones/es/aeat/ssii_1_1/fact/ws/SuministroFactEmitidas.wsdl'}
+        else:
+            return {'url': 'https://www2.agenciatributaria.gob.es/static_files/common/internet/dep/aplicaciones/es/aeat/ssii_1_1/fact/ws/SuministroFactRecibidas.wsdl'}
+
+    def _l10n_es_edi_web_service_bizkaia_vals(self, invoices):
+        if invoices[0].is_sale_document():
+            return {
+                'url': 'https://www.bizkaia.eus/ogasuna/sii/documentos/SuministroFactEmitidas.wsdl',
+                'test_url': 'https://pruapps.bizkaia.eus/SSII-FACT/ws/fe/SiiFactFEV1SOAP',
+            }
+        else:
+            return {
+                'url': 'https://www.bizkaia.eus/ogasuna/sii/documentos/SuministroFactRecibidas.wsdl',
+                'test_url': 'https://pruapps.bizkaia.eus/SSII-FACT/ws/fr/SiiFactFRV1SOAP',
+            }
+
+    def _l10n_es_edi_web_service_gipuzkoa_vals(self, invoices):
+        if invoices[0].is_sale_document():
+            return {
+                'url': 'https://egoitza.gipuzkoa.eus/ogasuna/sii/ficheros/v1.1/SuministroFactEmitidas.wsdl',
+                'test_url': 'https://sii-prep.egoitza.gipuzkoa.eus/JBS/HACI/SSII-FACT/ws/fe/SiiFactFEV1SOAP',
+            }
+        else:
+            return {
+                'url': 'https://egoitza.gipuzkoa.eus/ogasuna/sii/ficheros/v1.1/SuministroFactRecibidas.wsdl',
+                'test_url': 'https://sii-prep.egoitza.gipuzkoa.eus/JBS/HACI/SSII-FACT/ws/fr/SiiFactFRV1SOAP',
+            }
+
+    def _l10n_es_edi_call_web_service_sign(self, invoices, info_list):
+        company = invoices.company_id
+        # All are sharing the same value, see '_get_batch_key'.
+        csv_number = invoices.mapped('l10n_es_edi_csv')[0]
+
+        # === Decode the certificate ===
+
+        certificate = company.l10n_es_edi_certificate_id
+        pem_certificate, pem_private_key, certificate = certificate._decode_certificate()
+        with NamedTemporaryFile(suffix='.cert') as cert_file, NamedTemporaryFile(suffix='.pem') as pem_file:
+            cert_file.write(pem_certificate)
+            cert_file.flush()
+            pem_file.write(pem_private_key)
+            pem_file.flush()
+
+            # === Call the web service ===
+
+            class L10nEsHTTPAdapter(requests.adapters.HTTPAdapter):
+                """ An adapter to block DH ciphers which may not work for the tax agencies called"""
+                def init_poolmanager(self, *args, **kwargs):
+                    kwargs['ssl_context'] = create_urllib3_context(ciphers=EUSKADI_CIPHERS)
+                    return super().init_poolmanager(*args, **kwargs)
```

Comment:
For the record this technique was originally suggested [here](https://github.com/odoo/enterprise/pull/14181#discussion_r507584666) for AFIP servers because they are running in production with weak DH params. But this should not be copy/pasted elsewhere, /unless/:
- we can confirm that the *exact same problem happens* with this service (the traceback and error should be documented somewhere)
- we have no known way for contacting the authority for that service and getting them to solve the problem for good and for everyone (which is always better than this hack)

Reply:
@odony Are not they just using a better key automatically this way?  I think you mean this error when I just remove the ciphers=EUSKADI_CIPHERS: 

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/urllib3/contrib/pyopenssl.py", line 485, in wrap_socket
    cnx.do_handshake()
  File "/usr/lib/python3/dist-packages/OpenSSL/SSL.py", line 1915, in do_handshake
    self._raise_ssl_error(self._ssl, result)
  File "/usr/lib/python3/dist-packages/OpenSSL/SSL.py", line 1647, in _raise_ssl_error
    _raise_current_error()
  File "/usr/lib/python3/dist-packages/OpenSSL/_util.py", line 54, in exception_from_error_queue
    raise exception_type(errors)
OpenSSL.SSL.Error: [('SSL routines', 'tls_process_ske_dhe', 'dh key too small')]


urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='www2.agenciatributaria.gob.es', port=443): Max retries exceeded with url: /static_files/common/internet/dep/aplicaciones/es/aeat/ssii_1_1/fact/ws/SuministroFactEmitidas.wsdl (Caused by SSLError(SSLError("bad handshake: Error([('SSL routines', 'tls_process_ske_dhe', 'dh key too small')])")))